solls/LanguageServer.cpp

```diff
@@ -0,0 +1,679 @@
+		};
+
+		fprintf(stderr, " -> found reference %s at %d:%d .. %d:%d\n",
+			msg,
+			startLine, startColumn,
+			endLine, endColumn
+		);
+
+		auto highlight = lsp::Server::DocumentHighlight{};
+		highlight.range = locationRange;
+		highlight.kind = lsp::Server::DocumentHighlightKind::Text; // TODO: are you being read or written to?
+
+		m_result.emplace_back(highlight);
+	}
+
+	// TODO: MemberAccess
+
+	bool visitNode(ASTNode const& _node) override
+	{
+		if (&_node == &m_declaration)
+		{
+			if (auto const* decl = dynamic_cast<Declaration const*>(&_node))
+				addReference(decl->nameLocation(), "(visitNode)");
+			else
+				addReference(_node.location(), "(visitNode)");
+		}
+
+		return true;
+	}
+};
+
+} // }}} end helpers
+
+LanguageServer::LanguageServer(lsp::Transport& _client, Logger _logger):
+	lsp::Server(_client, std::move(_logger)),
+	m_vfs()
+{
+}
+
+void LanguageServer::shutdown()
+{
+	logInfo("LanguageServer: shutdown requested");
+}
+
+LanguageServer::InitializeResponse LanguageServer::initialize(
+	string _rootUri,
+	map<string, string> _settings,
+	Trace _trace,
+	vector<WorkspaceFolder> _workspaceFolders
+)
+{
+	(void) _trace; // TODO: debuglog based on this config
+	(void) _settings; // TODO: user settings (such as EVM version)
+
+#if !defined(NDEBUG)
+	ostringstream msg;
+	msg << "LanguageServer: rootUri : " << _rootUri << endl;
+	for (auto const& workspace: _workspaceFolders)
+		msg << "                workspace folder: " << workspace.name << "; " << workspace.uri << endl;
+#endif
+
+	if (boost::starts_with(_rootUri, "file:///"))
+	{
+		auto const fspath = boost::filesystem::path(_rootUri.substr(7));
+		m_basePath = fspath;
+		m_allowedDirectories.push_back(fspath);
+	}
+
+#if !defined(NDEBUG)
+	logMessage(msg.str());
+#endif
+
+	InitializeResponse hello{};
+	hello.serverName = "solc";
+	hello.serverVersion = "0.0.0"; // TODO: solidity version macro here
+	hello.supportsDefinition = true;
+	hello.supportsDocumentHighlight = true;
+	hello.supportsDocumentSync = true;
+	hello.supportsReferences = true;
+	hello.supportsHover = false; // TODO
+	return hello;
+}
+
+void LanguageServer::initialized()
+{
+	// NB: this means the client has finished initializing. Now we could maybe start sending
+	// events to the client.
+	logMessage("LanguageServer: Client initialized");
+}
+
+void LanguageServer::documentOpened(string const& _uri, string _languageId, int _documentVersion, std::string _contents)
+{
+	logMessage("LanguageServer: Opening document: " + _uri);
+
+	lsp::vfs::File const& file = m_vfs.insert(
+		_uri,
+		_languageId,
+		_documentVersion,
+		_contents
+	);
+
+	validate(file);
+}
+
+void LanguageServer::documentContentUpdated(string const& _uri, optional<int> _version, vector<DocumentChange> _changes)
+{
+	if (lsp::vfs::File* file = m_vfs.find(_uri); file != nullptr)
+	{
+		if (_version.has_value())
+			file->setVersion(_version.value());
+
+		for (DocumentChange const& change: _changes)
```

Comment:
I really don't see the need for this intermediate data structure.

Reply:
added to the todo list. I have to deal with that list of changes somewhere for sure, but i can move it away from solidity lsp into the lower layer, as solidity-lsp itself doesn't care about that except the fact that a file has changed (triggering a revalidation)